streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1940,107 +1940,117 @@ object ZChannel {
-                               .forkIn(childScope) *> latch.await
-                           }
-                         case MergeStrategy.BufferSliding =>
-                           pull.flatMap { channel =>
-                             val canceler = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
-                             val latch    = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
-                             for {
-                               size <- cancelers.size
-                               _    <- ZIO.when(size >= n0)(cancelers.take.flatMap(_.succeed(())))
-                               _    <- cancelers.offer(canceler)
-                               raceIOs =
-                                 ZIO.scopedWith { scope =>
-                                   (incoming >>> channel)
-                                     .toPullInAlt(scope)
-                                     .flatMap(evaluatePull(_).race(canceler.await.interruptible))
-                                 }
-                               _ <- permits
-                                      .withPermit(latch.succeed(()) *> raceIOs)
-                                      .interruptible
-                                      .forkIn(childScope)
-                               _ <- latch.await
-                             } yield ()
-                           }
-                       }
-        _ <- pullStrategy.forever
-               .catchAllCause(cause =>
-                 cause.failureOrCause match {
-                   case Left(_: Left[OutErr, OutDone]) => outgoing.offer(Exit.failCause(cause))
-                   case Left(x: Right[OutErr, OutDone]) =>
-                     permits.withPermits(n0)(ZIO.unit).interruptible *>
-                       lastDone.get.flatMap {
-                         case Some(lastDone) => outgoing.offer(Exit.fail(Right(f(lastDone, x.value))))
-                         case None           => outgoing.offer(Exit.fail(Right(x.value)))
-                       }
-                   case Right(cause) => outgoing.offer(Exit.failCause(cause.map(Left(_))))
-                 }
-               )
-               .raceFirst(awaitErrorSignal(childScope, fiberId)(errorSignal))
-               .forkIn(scope)
-      } yield {
-        lazy val consumer: ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone] =
-          unwrap[Env, Any, Any, Any, OutErr, OutElem, OutDone] {
-            outgoing.take.flatten.foldCause(
-              cause =>
-                cause.failureOrCause match {
-                  case Left(Left(outErr))   => ZChannel.fail(outErr)
-                  case Left(Right(outDone)) => ZChannel.succeedNow(outDone)
-                  case Right(cause)         => ZChannel.refailCause(cause)
-                },
-              outElem => ZChannel.write(outElem) *> consumer
+      (SingleProducerAsyncInput.make[InErr, InElem, InDone] <*> ZIO.fiberId).flatMap { case (input, fiberId) =>
```

Comment:
@kyri-petrou We could also make a unsafe version of this `.make` which would also remove 1 flatMap